sdk/include/opentelemetry/sdk/common/http_client.h

```diff
@@ -0,0 +1,129 @@
+#pragma once
+
+#include <chrono>
+#include <string>
+#include <vector>
+#include "opentelemetry/nostd/function_ref.h"
+#include "opentelemetry/version.h"
+
+// Usage Example
+//
+// struct ReponseHandler: public HttpResponseHandler {
+//      void OnHttpResponse(HttpResponse& res) noexcept override
+//      {
+//           if (res.IsSuccess())
+//              .. process response
+//      }
+//  };
+//
+//   HttpSessionManager sessionManager; // implementer can provide singleton implementation for it
+//   auto session = sessionManager.createHttpSession("localhost", 8000);
+//   auto request = session->CreateRequest();
+//   request->AddHeader(..);
+//   ResponseHandler res_handler;
+//   session->SendRequest(res_handler);
+//   session->FinishSession() // optionally in the end.
+//   ...shutdown
+//    sessionManager.FinishAllSessions()
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace common
+{
+namespace http
+{
+
+enum class HttpMethod
+{
+  HTTP_GET,
+  HTTP_POST,
+  HTTP_PUT,
+  HTTP_OPTIONS,
+  HTTP_HEAD,
+  HTTP_PATCH,
+  HTTP_DELETE
+};
+
+enum class HttpSessionState
+{
+  CREATED,  // session object is created
+  ONGOING,  // http(s) request is ongoing
+  FINISHED  // https(s) request is finished
+};
+
+using HttpHeader     = std::pair<std::string, std::string>;
+using HttpBody       = std::vector<uint8_t>;
+using HttpStatusCode = uint16_t;
+
+class HttpRequest
+{
+
+public:
+  virtual void setMethod(HttpMethod method) noexcept = 0;
+
+  virtual void setURI(const std::string &uri) noexcept = 0;
+
+  virtual void SetBody(HttpBody &body) noexcept = 0;
+
+  virtual void AddHeader(const HttpHeader &header) noexcept = 0;
+
+  virtual void ReplaceHeader(const HttpHeader &header) noexcept = 0;
+
+  virtual void SetTimeoutMs(std::chrono::milliseconds timeout_ms) noexcept = 0;
+};
+
+class HttpResponse
+{
+public:
+  virtual const HttpBody &GetBody() const noexcept = 0;
+
+  virtual bool GetNextHeader(nostd::function_ref<bool(HttpHeader)> callable) const noexcept = 0;
```

Comment:
How exactly does `GetNextHeader` work? Does it need to be called for each header? Or does it call `callable` for every header? Also, `bool(HttpHeader)` should probably rather be `bool(const &HttpHeader)`.

Reply:
It needs to be called for each header, something like this:
```
res.GetNextHeader([](opentelemetry::sdk::common::http::HttpHeader header) -> bool {
       std::cout << header.first << " " << header.second << "\n";
      return true;
});

Reply:
So, to print each header, I'd need to do this?
```c++
bool has_next;

do {
  has_next = res.GetNextHeader([](opentelemetry::sdk::common::http::HttpHeader header) -> bool {
    std::cout << header.first << " " << header.second << "\n";
    return true;
  });
} while (has_next);
```

Reply:
No, there need to be single `GetNextHeader()` call to iterate over all the headers. Please see the reference implementation of this method :
https://github.com/open-telemetry/opentelemetry-cpp/blob/ecb76205da27652a3668165fe9edf9b46c0261c0/exporters/include/opentelemtry/exporters/common/http/http_client.h#L71

Reply:
Have added this example as part of sample code in the header.

Reply:
I understand. Would a name like `ForEachHeader` be more clear to convey that intention?